### ISSUE TITLE
[Feedback] New infrastructure

### ISSUE THREAD

**Issue by funilrys** [Ultimate-Hosts-Blacklist/dev-center#32]
Hello, World!

Please report here every infrastructure/build/test issues (suspected or confirmed) from now.

**Note:** This is issue is only about something which is not normal with our new deployement logic. Everything else should be reported as a new issue.

Cheers,
Nissar

**Comment by funilrys** [comment]
First Issue:
```
Exception: Unable to get the content of 'https://raw.githubusercontent.com/Ultimate-Hosts-Blacklist/repository-structure/master/.travis.yml'.
```

It might be GitHub who blocks the machine as I'm able to access to https://raw.githubusercontent.com/Ultimate-Hosts-Blacklist/repository-structure/master/.travis.yml from my machine :thinking:

**Comment by dnmTX** [comment]
🤔

**Comment by funilrys** [comment]
Turn out it was an issue from my side. I forgot a return statement. That error did not happen in test phase because the `repository-structure` repository is the central one so it does not need to update `.travis.yml`.

Published to pip. Restarting all boxes who got the issue manually.

**Comment by funilrys** [comment]
Next one:
```
error: unknown option `procelain'
```

Also related to the update of `.travis.yml`. I might have use the wrong function. I'm looking for the fix.

**Comment by funilrys** [comment]
Turn out it is a typo. I used `--procelain` instead of `--porcelain`.

Published to pip. Restarting all boxes who got the issue manually.

**Comment by funilrys** [comment]
Next one:
```
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/bin/ultimate-hosts-blacklist-input-repo-updater", line 10, in <module>
    sys.exit(_command_line())
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/ultimate_hosts_blacklist/input_repo_updater/__init__.py", line 93, in _command_line
    logging_level=logging_level, multiprocessing=arguments.multiprocess
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/ultimate_hosts_blacklist/input_repo_updater/core.py", line 840, in process
    self.__process_multiprocess(to_test, end_time)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/ultimate_hosts_blacklist/input_repo_updater/core.py", line 604, in __process_multiprocess
    continue_data = List(continue_data).merge(data, strict=False)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/ultimate_hosts_blacklist/helpers/list.py", line 144, in merge
    result.append(element)
AttributeError: 'dict' object has no attribute 'append'
```

This time it's just before pushing everything to the repository. It is related to the new continue subsystem which is specially implemented for our infrastructure. I'm looking at the code.

**Comment by funilrys** [comment]
Turns out It was a naming issue. Indeed I used `List(xx).merge(xxx)` instead of `Dict(xx).merge(xxx)`.
Both `List()` and `Dict()` classes are part of our helpers.

Published to pip. Manually restarting all boxes who got the issue.

**Comment by funilrys** [comment]
@dnmTX I'm going to sleep but a brunch of repositories already finished their test and generation.

**Comment by dnmTX** [comment]
@funilrys  looks like https://github.com/Ultimate-Hosts-Blacklist/Spotify-Ad-free doesn't exist anymore and the build failed because of it( some kind of a DMCA takedown).

Also i see some of the builds failed for other reasons and on some that passed the `volatile.list` and `whitelisted.list` are not updated. Please check tomorrow when you have time.Good night.

**Comment by dnmTX** [comment]
just checked the `volatile.list` in [lightswitch05's repo](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended) and domains(active for sure) from *.doubleclick.net were not added(some are there but many are missing).For example:
`1044889.fls.doubleclick.net`
`1063127.fls.doubleclick.net`
`1092360.fls.doubleclick.net`
`1095311.fls.doubleclick.net`
`1106306.fls.doubleclick.net`
`1119706.fls.doubleclick.net`
`1181183.fls.doubleclick.net`
`1268437.fls.doubleclick.net`
`1272738.fls.doubleclick.net`

**Comment by funilrys** [comment]
`volatile` is priority. I'm checking that right now.

**Comment by funilrys** [comment]
@dnmTX @mitchellkrogza 
Just a feedback on the issue regarding the `volatile`. My last commit should fix it in the future.

But when looking at logs I found out that the DNS server under the container do not resolve correctly for some domains. Which caused PyFunceble to return `INACTIVE`.

For about 2 hours I started to write the code which will let us use different DNS server than the one installed. The code is almost ready. Will be out in the coming hours.

I need some fresh air so catch you later.

**Comment by dnmTX** [comment]
Great. Thanks @funilrys  👍 
Just curious,which DNS server was implemented initially?

**Comment by funilrys** [comment]
The OS installed one so in our case under Travis CI ([cf](https://travis-ci.org/Ultimate-Hosts-Blacklist/repository-structure/builds/531101135#L240-L244)): 
```
$ cat /etc/resolv.conf
# Dynamic resolv.conf(5) file for glibc resolver(3) generated by resolvconf(8)
#     DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
nameserver 169.254.169.254
search c.eco-emissary-99515.internal google.internal
```

**Comment by dnmTX** [comment]
Yeah....that IP....very suspicious.Basic check showed that some commercial domains are associate with it.
Better change it.Even Google(8.8.8.8) is better then that.

**Comment by funilrys** [comment]
Yeah, that's also my conclusion when I saw that. Unfortunately, Travis CI don't allow any changes to that file ([cf](https://travis-ci.org/Ultimate-Hosts-Blacklist/repository-structure/builds/531101135#L247-L249)) ... 

That's why I had to edit PyFunceble for the job  ... We lose a bit in performance but it's balanced with the new multiprocess usage which allows us to simultaneously test as many as possible before 15 minutes is reached.

```
$ sudo printf "nameserver 1.1.1.1\nnameserver 1.0.0.1\n" > /etc/resolv.conf
/home/travis/.travis/functions: line 104: /etc/resolv.conf: Permission denied
```

**Comment by dnmTX** [comment]
I saw that and also i think there is a typo: `nnameserver 1.0.0.1\n"` double n.

**Comment by funilrys** [comment]
No no typo @dnmTX:
```
$ printf "nameserver 1.1.1.1\nnameserver 1.0.0.1\n"
nameserver 1.1.1.1
nameserver 1.0.0.1
```

2.x.x is getting stable thanks to what we both do here so when we are done with all the issue 2.x.x will be the most stable version of PyFunceble ever :wink:

**Comment by dnmTX** [comment]
I'll keep monitoring and will report after next filtering is done.

**Comment by funilrys** [comment]
@dnmTX Following are fixed. Test result at https://github.com/Ultimate-Hosts-Blacklist/repository-structure/commit/16752ff83a86a5524a6977af58d0159bab337b49

* https://github.com/Ultimate-Hosts-Blacklist/dev-center/issues/32#issuecomment-491475932
* https://github.com/Ultimate-Hosts-Blacklist/dev-center/issues/32#issuecomment-491475932

Now let's focus on those still red one :smile_cat:

**Comment by dnmTX** [comment]
Ok,let me summarize:
1. Missing *.doubleclick.net domains -FIXED (DNS resolver's fault)
2. `volatile.list` `whitelisted.list` not updating after filtering -FIXED
3. `www.` not applied in `domains.list` before initial filtering -if this was the case i'm still waiting on status for this one

**Comment by funilrys** [comment]
Manually restarted `lightswitch05_hosts_ads-and-tracking-extended`

**Comment by funilrys** [comment]
@dnmTX `www.` and vice-versa are now tested and generated on the fly. No more pre-initialization saved into `domains.list`.

Which means (This is an example for one line but we do that for as most line as possible - and simultaneously):
* Download upstream into `domains.list`
* Read lines of `domains.list` ==> `google.com` 
* Generate complements ==> `[www.google.com, google.com]`
* Test every element of the list ==> `www.google.com` then `google.com` 
* Read next line of `domains.list`

**Comment by dnmTX** [comment]
Oh,i see.Well...that's a great improvement(i can scratch that from the list then).Just...again....to make sure...subdomains are skipped or....?

**Comment by funilrys** [comment]
Yes subdomains are skipped :)

**Comment by dnmTX** [comment]
Hehehe...just saw this in my router's log.It's fro when i was checking the domains that were associated with 
Travis Cl's DNS resolver:
`May 11 12:25:23 DD-WRT daemon.warn dnsmasq[1315]: possible DNS-rebind attack detected: maisonmargielaparfums.com`

**Comment by funilrys** [comment]
Another phishing domain 🤔

**Comment by funilrys** [comment]
Nope just some bad boyz :disappointed:

**Comment by dnmTX** [comment]
@funilrys  how's it going with the repairs? I just checked lightswitch05's repo and it looks good as far as i can tell. Just checking on the progress. 😃

**Comment by funilrys** [comment]
Good @dnmTX :smile_cat: I restarted every red flagged this morning and they all passed successfully 🙂 
There was some exception regarding the new DNS lookup logic of PyFunceble but it was quickly fixed. Unfortunally those exceptions were not flagged as the exceptions were not passed to the parent process :smile: 

So right now I'm locally testing a way to completly stop PyFunceble when issue happen for the new `--multiprocess` argument. Once I found a proper solution I will implement that that our updater 🙂 

Why twice the job ? Because 2.0.0 will have the same problem as our updated once distributed :joy_cat: 

Indeed we (here) include the PyFunceble 2.0.0 API and do our business logic around its output when everybody else will use the `--multiprocess`  argument from the CLI once released 🙂 So same problem to different scale and target :joy_cat:

**Comment by funilrys** [comment]
Oups my bad :joy_cat:

**Comment by dnmTX** [comment]
@funilrys  couple builds failed so far,most likely is connected to the DNS resolving(as far as i can tell).
Check it out when you have the time.Otherwise everything else looks good 👍

**Comment by funilrys** [comment]
Should be fixed. Restarting all flagged one.

**Comment by dnmTX** [comment]
@funilrys  there are still some builds that are failing: [ZeroDot1](https://github.com/Ultimate-Hosts-Blacklist/ZeroDot1_CoinBlockerLists) for example among few others.Can you please take a look.Thank you.

**Comment by dnmTX** [comment]
@funilrys  builds are still failing after the most recent fixes and there are some that it says passing but has been filtering since yesterday(i'm guessing it shouldn't take that long).

**Comment by funilrys** [comment]
Hi @dnmTX I did not forgot you, just developing a way to get rid of all the databases in order to gain in time and space. (At least for near the future).

I'll restart everything in the next hours.

Cheers,
Nissar

**Comment by funilrys** [comment]
Restart of all repository started.

**Comment by funilrys** [comment]
Restart of all repositories finished.

**Comment by funilrys** [comment]
Another restart session will be run later as I forgot to first sync the `repository-structure` repository.

**Comment by funilrys** [comment]
For the record:

Due to changes in the PyFunceble API, I had to rewrite some section of the updater.

**Comment by funilrys** [comment]
Restart of all repositories started.

**Comment by funilrys** [comment]
Everything stopped because I might have reached GitHub API token limit.

**Comment by funilrys** [comment]
Restart of all repositories finished.

**Comment by dnmTX** [comment]
@funilrys  the restart(from what i can see) didn't do much and as of late i've been busy and didn't and will not have the time to monitor,report and most importantly pushing you to do the fixes.It's been almost 3 weeks without any of my chosen lists from here got any updates and i'd really like whatever free time i'll have from now on to spend on my PC is to to at least have some updated protection.
I know you busy too and this is not me blaming you or whatever,i just need to know what is your plans to bring the filtering back to where it was(super stable) cause if it's not going to be soon i'd rather switch to the original lists and not to worry about it at least for a while.

**Comment by funilrys** [comment]
Hi @dnmTX, 
Sorry for my bad behavior in those last 3 weeks. I was just too focused on improving, testing, reviewing, fixing PyFunceble so that in the future it will be easier to maintain and understand by my peer. Fortunately (and thanks to @mitchellkrogza), we did a great job this week with @mitchellkrogza around that objective and no more major issue was found in the last 48 hours.

About the infrastructure, a patch which reviews the way we launch multiple tests at the time has been pushed. I'll monitor everything right now and until 01:30 (Berlin time). The objective for me is to get a list of traceback with the new patch so that I can take a big time of my Sunday to find solutions.

Everything was restarted about 10 minutes ago, it should be finished in the coming minutes.

Everything should come back and running the next week. If multiprocessing is not fixed, I'll drop it so that everybody can globally get new datasets.

Thanks for your understanding, your support and time.

Cheers,
Nissar

**Comment by funilrys** [comment]
Just feedback before going to bed: No infrastructure issue from the last comment to now.

There are still 26/60+ repositories which are under test but since my patch, I did not get any failed nor error tests.

I took the time to search for traceback into the logs of all the one who finished, and I wasn't able to find some. 

Will review the rest later in the morning but for me, the patch was the right one!

Cheers,
Nissar

**Comment by dnmTX** [comment]
@funilrys  great 👍 .Looking forward to turn my cron jobs back on. When i have time(during the week) will try to monitor it and will report back(if anything) in the upcoming weekend. Thank you for the hard work you doing,crossing fingers tomorrow's tests to go smoothly. Good night.

**Comment by dnmTX** [comment]
@funilrys  how's it going with the fixes? Can you please give some update on the progress from your end.
Just to note here that even though it says that the builds are passing none of the lists has been updated,it's like it's going into a loop all the time.

**Comment by funilrys** [comment]
Hi @dnmTX,

For me, everything works back! If there are no changes into `clean.list` or `volatile.list` it's because nothing changed.  And no after checking, there is no loop!

I'm going to include the times for the future into `info.json` so you can work with it but on my side, everything works. 

The only difference is in order to avoid an inaccurate test result overtime because we test as many as possible - and some networks might not like it - I reduced the execution time of one build from ~15 minutes to ~5 minutes. That way it also allows us to test more input sources in the same hour 😊

Cheers,
Nissar

**Comment by dnmTX** [comment]
ok @funilrys  i'll give you couple examples:

1. [lightswitch05](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended) it says build error,the `domains.list` was last updated 8 days ago but as you can see from the original lists there was [commit](https://github.com/lightswitch05/hosts/commit/3b446d4259877264c49fd37677b519cbba9e8cc1) 4 days ago.
Neither `volatile.list`,`clean.list` or `whitelisted.list` were updated for 25 days so far which,i'd assume is when all stopped working.

2. [ZeroDot1](https://github.com/Ultimate-Hosts-Blacklist/ZeroDot1_CoinBlockerLists) `domains.list` last updated 10 days ago,[last commit](https://gitlab.com/ZeroDot1/CoinBlockerLists/commit/5e730bf8810f2b96efccddce96929ca8a7ac22be) 3 days ago.Neither of the other lists were updated for 28 days.

There are others.Sorry but,seeing this i can't convince myself that the filtering is anywhere near working properly.

**Comment by funilrys** [comment]
@dnmTX:

* ZeroDot1 still under test the last 10 days ... 
* lightswitch05 ... Travis CI issue ...

**Comment by funilrys** [comment]
You're right those are not filtering but infrastructure issue. Will look into a way to speed up the test overall in our infrastructure along with a way to restart if a new version is present ... From now it's doesn't have anything to do with PyFunceble ... Just the way our infrastructure live and work...

**Comment by dnmTX** [comment]
> You're right those are not filtering but infrastructure issue.

Well,in my defense if the infrastructure is broken anyone would assume that there is something wrong with the filtering but thank's for acknowledging that there is a issue and please,keep me posted this time  on what did you fixed.This way i'll monitor more closely and give you confirmation.
Now...let's get the ball rolling and fix that damn infrastructure once and for all 😄

**Comment by funilrys** [comment]
Time increased from 5 to 10 minutes.

**Comment by funilrys** [comment]
We now test 25 subjects at the same time instead of the overloading "as much as possible".

**Comment by dnmTX** [comment]
Also reminder to ADD the new,improved and unified [LIST](https://github.com/smed79/blacklist/blob/master/hosts.txt) from our friend @smed79  😉 👍  and remove all the rest which contain smed79 in them as they're not needed anymore.

**Comment by dnmTX** [comment]
....and after...approximately...28 days i was finally able to update all of my lists.
I call this progress 😅 
Thank you @funilrys  ❤️ ❤️ ❤️  👍 💯

**Comment by funilrys** [comment]
@dnmTX I lost my automation tool for repository creation (Yeah I made the stupid error to not create a private repo for it :sob:) so everything around repo creation/deletion will have to wait a bit!

**Comment by dnmTX** [comment]
@funilrys  no rush,at least we know that it's in your **TO DO LIST** and thanks for the update 😉

**Comment by dnmTX** [comment]
@funilrys looks like the [INVALID](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended/blob/master/output/domains/INVALID/list) folder doesn't refresh after each filtering and there are some duplicates there and when comparing it to the latest downloaded [domains.list](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended/blob/master/domains.list) those domains are not even present there anymore:
![lightswitch05](https://user-images.githubusercontent.com/34774426/60229208-c02a7f80-985a-11e9-81cf-8cc321a78529.png)

Please take a look when you have time. Thanks 👍

**Comment by funilrys** [comment]
Added to the workflow :+1:

**Comment by funilrys** [comment]
Hello there,

just a short update, I'm almost done with the repository creation logic/tool 🙂 I'll start with @smed79's list later today (Berlin Time). All other issues which are in my workflow are in my internal queue. 

Cheers,
Nissar

**Comment by funilrys** [comment]
Just created @smed79 :+1:

**Comment by funilrys** [comment]
Waiting for push confirmation from the test with PyFunceble so I can delete the others :)

**Comment by funilrys** [comment]
> @funilrys looks like the [INVALID](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended/blob/master/output/domains/INVALID/list) folder doesn't refresh after each filtering and there are some duplicates. Also when comparing it to the latest downloaded [domains.list](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended/blob/master/domains.list) those (invalid) domains are not even present there anymore:
> 
> ![lightswitch05](https://user-images.githubusercontent.com/34774426/60229208-c02a7f80-985a-11e9-81cf-8cc321a78529.png)
> 
> Please take a look when you have time. Thanks

@dnmTX  About that issue I'll start working on that next. It should not have influences in the outputs as those are re-test outputs but I do understand the confusion it creates.

**Comment by funilrys** [comment]
@smed79 @dnmTX Do you have a list of smed79_* repository to delete?

**Comment by dnmTX** [comment]
> @smed79 @dnmTX Do you have a list of smed79_* repository to delete?

1. https://github.com/Ultimate-Hosts-Blacklist/smed79_hilltopads_adservers
2. https://github.com/Ultimate-Hosts-Blacklist/smed79_admeasures_adservers
3. https://github.com/Ultimate-Hosts-Blacklist/smed79_getadmiral.com
4. https://github.com/Ultimate-Hosts-Blacklist/smed79_popads_adservers
5. https://github.com/Ultimate-Hosts-Blacklist/smed79_propellerads_adservers
6. https://github.com/Ultimate-Hosts-Blacklist/smed79_blocklist_facebook  **(not sure about this one)**
7. https://github.com/Ultimate-Hosts-Blacklist/smed79_parkingcrew.com
8. https://github.com/Ultimate-Hosts-Blacklist/smed79_uponit.com_adservers
9. https://github.com/Ultimate-Hosts-Blacklist/smed79_assorted

@smed79  can you confirm the list above please.Thanks.

@funilrys  thank you for adding the new @smed79's unified lists.Very much appreciated and thanks for fixing the issue with the invalid domains 👍

**Comment by dnmTX** [comment]
@funilrys  there is something wrong with the infrastructure again. Some builds showing error,on others the lists are not updating,even though it shows that it passed.Can you please take a look.Thanks.

**Comment by smed79** [comment]
@funilrys 

> Do you have a list of smed79_* repository to delete?

keep only :
- [smed79_blacklist](https://github.com/Ultimate-Hosts-Blacklist/smed79_blacklist)
- [smed79_blocklist_facebook](https://github.com/Ultimate-Hosts-Blacklist/smed79_blocklist_facebook)

fyi, in the day that follows i am planning to upload a [nsfw list](https://github.com/smed79/blacklist/blob/master/extra/nsfw.txt).

**Comment by funilrys** [comment]
@smed79 ==> :heavy_check_mark: + [:checkered_flag:](https://github.com/Ultimate-Hosts-Blacklist/smed79_blacklist_nsfw)(it's a link :wink:)

@dnmTX Was a Travis CI. I'm planning a full restart this night (Berlin Time).

**Comment by dnmTX** [comment]
@funilrys  still not fixed. Please when you have time do a deeper assessment and hopefully get it fixed.
Thanks a bunch 👍

**Comment by dnmTX** [comment]
@funilrys where are all those [INVALID DOMAINS](https://github.com/Ultimate-Hosts-Blacklist/lightswitch05_hosts_ads-and-tracking-extended/blob/master/output/hosts/INVALID/hosts) coming from if they don't even exist in the same form in the original lists?
This is definitely some parsing bug that needs to be looked at.
![tracking-extended](https://user-images.githubusercontent.com/34774426/63231214-250cb100-c1de-11e9-925c-13685b4514ef.png)

**Comment by funilrys** [comment]
@dnmTX They come from the inactive database aka `inactive-db.json`. I fixed that on the mainstream.

Indeed, from now, if a subject is listed into the inactive database and the status is still the same (also `INACTIVE` or `INVALID`) it's not included in the results.

For our use case (our infrastructure), https://github.com/funilrys/PyFunceble/commit/670921776524ccf74a4d53d6ffa3d6a4a74f4447 make the mainstream patch available because we use the API here and not the CLI.

**Comment by funilrys** [comment]
Duplicated entries ==> in progress

**Comment by dnmTX** [comment]
@funilrys  some builds are failing:
https://github.com/Ultimate-Hosts-Blacklist/ZeroDot1_CoinBlockerLists
https://github.com/Ultimate-Hosts-Blacklist/anudeepND_blacklist_ad_server
And this one is not updating the lists:
https://github.com/Ultimate-Hosts-Blacklist/quidsup_malicious-sites

Please take a look when you have time. Thank you 👍

**Comment by dnmTX** [comment]
@funilrys  as of late builds are passing but none of the lists got updated,last update on most of them is showing 8 days ago.Hopefully you'll have time to fix it.Thank you 👍

**Comment by funilrys** [comment]
Hi @dnmTX, I hope that everything is OK on your side!

Since last weekend we released our monitoring tool which I wrote for the past month at hosts . Ubuntu 101 . Co . za / app / infra (remove spaces)

I still have to adjust it but if it's correct. Indeed, as we launch test more often, sometime the results won't change. Which come to what you're discribing ☺️


Have a nice day!

Cheers
Nissar

**Comment by dnmTX** [comment]
>  Indeed, as we launch test more often, sometime the results won't change. Which come to what you're discribing

@funilrys  you know,i've been monitoring certain lists for a long time now and never before i have seen none of the lists to not be updated for a period of 12-14 days. So far,every time when test i launched the results are different(no exception).This is not normal,and needs to be looked at.

**Comment by funilrys** [comment]
@dnmTX, you were (one again right). After testing, and reading the last 10 days of logs of our infrastructure launcher running in my raspberry pi, I finally found it.

The patch (PyFunceble) is on its way. I created a git-hook on my machine to check that it still works before committing. 

Thanks again for everything!

**Comment by funilrys** [comment]
Infrastructure restart (from the beginning of each list) ordered. Should start in the next minutes/hour.

Time to go to bed. 
Catch up later!

**Comment by funilrys** [comment]
Closing, Please report to the weekly issues.
Thanks again for everything!
